### PR TITLE
Kdesktop 754 Snapshot invalidated during compute FS operation phase

### PR DESCRIPTION
**Pull request by ChristopheLarchier**

### CODE REVIEW

File: src/libsyncengine/syncpal/operationprocessor.cpp [hunk 1]
```diff
@@ -32,9 +32,10 @@ bool OperationProcessor::isPseudoConflict(std::shared_ptr<Node> node, std::share
     }
 
     std::shared_ptr<Snapshot> snapshot =
-        (node->side() == ReplicaSide::ReplicaSideLocal ? _syncPal->_localSnapshot : _syncPal->_remoteSnapshot);
+        (node->side() == ReplicaSide::ReplicaSideLocal ? _syncPal->_localSnapshotCopy : _syncPal->_remoteSnapshotCopy);
```
Comment (luc-guyot-infomaniak) [inline src/libsyncengine/syncpal/operationprocessor.cpp:35]: Suggestion: create a helper function that retrieves the snapshot copy given the replica side as input (the same operation is done several times below).

File: src/libsyncengine/syncpal/syncpal.cpp [hunk 2]
```diff
@@ -1022,8 +1032,8 @@ ExitCode SyncPal::updateSyncNode(SyncNodeType syncNodeType) {
 
     auto nodeIdIt = nodeIdSet.begin();
     while (nodeIdIt != nodeIdSet.end()) {
-        bool ok = syncNodeType == SyncNodeTypeTmpLocalBlacklist ? _localSnapshot->exists(*nodeIdIt)
-                                                                : _remoteSnapshot->exists(*nodeIdIt);
+        bool ok = syncNodeType == SyncNodeTypeTmpLocalBlacklist ? _localSnapshotCopy->exists(*nodeIdIt)
```
Comment (luc-guyot-infomaniak) [inline src/libsyncengine/syncpal/syncpal.cpp:1035]:
```suggestion
      const  bool ok = syncNodeType == SyncNodeTypeTmpLocalBlacklist ? _localSnapshotCopy->exists(*nodeIdIt)
```

**Comment by luc-guyot-infomaniak** [comment]
Do the change match the pull-request title?

File: src/libsyncengine/syncpal/operationprocessor.cpp (hunk 1 above)
Comment (ChristopheLarchier) [reply src/libsyncengine/syncpal/operationprocessor.cpp:35]: Done

File: src/libsyncengine/syncpal/syncpal.cpp (hunk 2 above)
Comment (ChristopheLarchier) [reply src/libsyncengine/syncpal/syncpal.cpp:937]: Done

File: src/libsyncengine/syncpal/syncpal.h [hunk 3]
```diff
@@ -279,6 +281,8 @@ class SYNCENGINE_EXPORT SyncPal : public std::enable_shared_from_this<SyncPal> {
         std::shared_ptr<bool> _interruptSync;
         std::shared_ptr<Snapshot> _localSnapshot;
         std::shared_ptr<Snapshot> _remoteSnapshot;
+        std::shared_ptr<Snapshot> _localSnapshotCopy;
```
[the remark this answers is not in the document]
Comment (ChristopheLarchier) [reply src/libsyncengine/syncpal/syncpal.h]: Done

File: src/libsyncengine/syncpal/syncpal.h [hunk 4]
```diff
@@ -223,7 +223,9 @@ class SYNCENGINE_EXPORT SyncPal : public std::enable_shared_from_this<SyncPal> {
         void refreshTmpBlacklist();
         void removeItemFromTmpBlacklist(const NodeId &nodeId, ReplicaSide side);
 
-        std::shared_ptr<UpdateTree> getUpdateTree(ReplicaSide side) { return side == ReplicaSideLocal ? _localUpdateTree : _remoteUpdateTree; }
+		std::shared_ptr<UpdateTree> getUpdateTree(ReplicaSide side) { return side == ReplicaSideLocal ? _localUpdateTree : _remoteUpdateTree; }
+
+        void copySnapshots();
```
[the remark this answers is not in the document]
Comment (ChristopheLarchier) [reply src/libsyncengine/syncpal/syncpal.h:229]: Done

**Pull request merged**